File: Medium/430. Flatten a Multilevel Doubly Linked List/FlattenAMultilevelDoublyLinkedList.py
'''
LeetCode Medium: 430. Flatten a Multilevel Doubly Linked List
https://leetcode.com/problems/flatten-a-multilevel-doubly-linked-list/

Check FlattenAMultilevelDoublyLinkedList.cpp for brief solution explanation.
'''
class Solution:
    def traverse(self, head, ans):
        if head==None:
            return
        ans.append(head.val)
        self.traverse(head.child, ans)
        self.traverse(head.next, ans)

    def flatten(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head==None:
            return None
        ans = []
        self.traverse(head, ans)
        root = Node()
        cur = root
        prev = None
        for i in range(len(ans)):
            cur.val = ans[i]
            cur.prev = prev
            if i!=len(ans)-1:
                cur.next = Node()
            cur.child = None
            prev = cur
            cur = cur.next
        return root

File: Medium/430. Flatten a Multilevel Doubly Linked List/FlattenAMultilevelDoublyLinkedList.java
/*
LeetCode Medium: 430. Flatten a Multilevel Doubly Linked List
https://leetcode.com/problems/flatten-a-multilevel-doubly-linked-list/

Check FlattenAMultilevelDoublyLinkedList.cpp for brief solution explanation.
*/
class Solution {
    public void traverse(Node head, ArrayList<Integer>ans){
        if(head==null){
            return;
        }
        ans.add(head.val);
        traverse(head.child, ans);
        traverse(head.next, ans);
    }
    public Node flatten(Node head) {
        if(head==null){
            return null;
        }
        ArrayList<Integer>ans = new ArrayList<Integer>();
        traverse(head, ans);
        Node root = new Node();
        Node cur = root;
        Node prev = null;
        for(int i = 0;i<ans.size();i++){
            cur.val = ans.get(i);
            cur.prev = prev;
            if(i!=ans.size()-1){
                cur.next = new Node();
            }
            cur.child = null;
            prev = cur;
            cur = cur.next;
        }
        return root;
    }
}